### ISSUE TITLE
gRPC-JS client hangs when target server restarts

### ISSUE THREAD

**Issue by sheinbergon** [grpc/grpc-node#2785]
### Problem description

gRPC js client halts on server disconnections

### Reproduction steps

We have a server (server API is gRPC, using malijs), let's call it **A** which calls a gRPC target proxied by an NLB and nginx gRPC ingress, let's call it **B**. If we go ahead and restart the target instances of **B**, The client might sometime hang, resulting in the application **A** failing to function. All flows calling to A that will result in **A** calling **B** will hang

The moment we restart **A**, it becomes functioning, meaning it's a client state issue.

### Environment

 - Debian 11/12 based containers 
 - Node version: 18.20.x and 22.2.x
 - NVM based installation
 - gRPC-JS 1.19.15
-  Client options: `{ 'grpc.keepalive_time_ms': 10_000, 'grpc.max_receive_message_length': 64 * 1024 * 1024 }`

### Additional context

We've activated gRPC tracing on server **A**

here's what a normal server selection trace looks like: 

```
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | subchannel | (157) 10.41.10.107:443 READY -> IDLE
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | subchannel_refcount | (157) 10.41.10.107:443 refcount 2 -> 1
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | pick_first | READY -> IDLE
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | resolving_load_balancer | dns:service-b.internal.com:443 READY -> IDLE
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | connectivity_state | (2) dns:service-b.internal.com:443 READY -> IDLE
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | dns_resolver | Looking up DNS hostname service-b.internal.com
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | resolving_load_balancer | dns:service-b.internal.com:443 IDLE -> CONNECTING
D 2024-07-04T09:37:26.487Z | v1.9.15 29 | connectivity_state | (2) dns:service-b.internal.com:443 IDLE -> CONNECTING
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | dns_resolver | Resolved addresses for target dns:service-b.internal.com:443: [10.41.33.32:443,10.41.22.168:443,10.41.10.107:443]
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel | (202) 10.41.33.32:443 Subchannel constructed with options {}
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel_refcount | (202) 10.41.33.32:443 refcount 0 -> 1
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel | (203) 10.41.22.168:443 Subchannel constructed with options {}
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel_refcount | (203) 10.41.22.168:443 refcount 0 -> 1
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel_refcount | (202) 10.41.33.32:443 refcount 1 -> 2
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel_refcount | (203) 10.41.22.168:443 refcount 1 -> 2
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel_refcount | (157) 10.41.10.107:443 refcount 1 -> 2
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | pick_first | Start connecting to subchannel with address 10.41.33.32:443
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | pick_first | IDLE -> CONNECTING
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | resolving_load_balancer | dns:service-b.internal.com:443 CONNECTING -> CONNECTING
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | connectivity_state | (2) dns:service-b.internal.com:443 CONNECTING -> CONNECTING
D 2024-07-04T09:37:26.492Z | v1.9.15 29 | subchannel | (202) 10.41.33.32:443 IDLE -> CONNECTING
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | subchannel | (202) 10.41.33.32:443 CONNECTING -> READY
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | pick_first | Pick subchannel with address 10.41.33.32:443
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | subchannel_refcount | (202) 10.41.33.32:443 refcount 2 -> 3
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | subchannel_refcount | (202) 10.41.33.32:443 refcount 3 -> 2
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | subchannel_refcount | (203) 10.41.22.168:443 refcount 2 -> 1
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | subchannel_refcount | (157) 10.41.10.107:443 refcount 2 -> 1
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | pick_first | CONNECTING -> READY
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | resolving_load_balancer | dns:service-b.internal.com:443 CONNECTING -> READY
D 2024-07-04T09:37:26.499Z | v1.9.15 29 | connectivity_state | (2) dns:service-b.internal.com:443 CONNECTING -> READY
D 2024-07-04T09:37:30.637Z | v1.9.15 29 | subchannel_refcount | (157) 10.41.10.107:443 refcount 1 -> 0
```

As expected, connections enter into `TRANSIENT_FAILURE` state once I kill the target pods *B*. When they become responsive again, We get these messages:

```
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (299) 10.41.22.168:443 CONNECTING -> TRANSIENT_FAILURE
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (299) 10.41.22.168:443 TRANSIENT_FAILURE -> IDLE
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (299) 10.41.22.168:443 IDLE -> CONNECTING
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (300) 10.41.33.32:443 CONNECTING -> TRANSIENT_FAILURE
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (300) 10.41.33.32:443 TRANSIENT_FAILURE -> IDLE
D 2024-07-04T09:47:31.084Z | v1.9.15 29 | subchannel | (300) 10.41.33.32:443 IDLE -> CONNECTING
D 2024-07-04T09:47:31.093Z | v1.9.15 29 | subchannel | (300) 10.41.33.32:443 CONNECTING -> READY
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | pick_first | Pick subchannel with address 10.41.33.32:443
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | subchannel_refcount | (300) 10.41.33.32:443 refcount 2 -> 3
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | subchannel_refcount | (275) 10.41.10.107:443 refcount 2 -> 1
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | subchannel_refcount | (300) 10.41.33.32:443 refcount 3 -> 2
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | subchannel_refcount | (299) 10.41.22.168:443 refcount 2 -> 1
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | pick_first | TRANSIENT_FAILURE -> READY
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | resolving_load_balancer | dns:service-b.internal.com:443 TRANSIENT_FAILURE -> READY
D 2024-07-04T09:47:31.094Z | v1.9.15 29 | connectivity_state | (2) dns:service-b.internal.com:443 TRANSIENT_FAILURE -> READY
D 2024-07-04T09:47:31.109Z | v1.9.15 29 | subchannel | (299) 10.41.22.168:443 CONNECTING -> READY
```

And from that moment on, we stop seeing these messages altogether. As though the driver becomes stuck.

We still see these traces
```
D 2024-07-04T09:43:57.000Z | v1.9.15 29 | channel | (2) dns:service-b.internal.com:443 createResolvingCall [272] method="...", deadline=Infinity
D 2024-07-04T09:43:57.001Z | v1.9.15 29 | channel | (2) dns:service-b.internal.com:443 createRetryingCall [273] method="..."
D 2024-07-04T09:43:57.001Z | v1.9.15 29 | channel | (2) dns:service-b.internal.com:443 createLoadBalancingCall [274] method="..."
```

but no request is actuall being sent, the client is completely frozen. Any help will be much obliged

**Comment by murgatroid99** [comment]
It looks like you enabled only connection management trace logs, but we will also need trace logs for individual requests to determine what is happening with requests that is causing them not to be sent. It's probably simplest to just use `GRPC_TRACE=all` to get all the logs we might need.

**Comment by aikoven** [comment]
We hit what looks like the same failure mode in production and managed to reduce
it to a ~25-line reproduction. Posting it here in case it helps unblock this
issue, since it's marked `requires reporter action`.

**TL;DR:** a connection attempt that completes at the TCP level but never
receives the server's HTTP/2 SETTINGS frame hangs forever. The channel stays in
`CONNECTING` indefinitely, no backoff retry is ever scheduled, and calls without
a deadline queue up silently — exactly the "calls are created but nothing is
sent" symptom described in this issue.

## Root cause

`Http2SubchannelConnector.createSession()` waits for the `remoteSettings` event
with no timeout. It only attaches:

```js
session.once('remoteSettings', () => {
  /* resolve */
});
session.once('close', closeHandler);
session.once('error', errorHandler);
secureConnectResult.socket.once('close', closeHandler);
secureConnectResult.socket.once('error', errorHandler);
```

If the peer accepts the TCP connection (and completes the TLS handshake) but
never sends SETTINGS, none of these fire. The returned promise never settles, so
`Subchannel.startConnecting()` never transitions out of `CONNECTING` and the
backoff timer — which is only armed on connection _failure_ — is never started.

This is also a spec deviation: per
[`connection-backoff.md`](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md),
`TryConnect` should be called with a deadline of at least `MIN_CONNECT_TIMEOUT`
(20 s) and fail into backoff when that expires. C-core and grpc-java do this;
grpc-js has no equivalent of `MIN_CONNECT_TIMEOUT`.

Keepalive doesn't help, because it only applies to an already-established
transport — there is no transport yet.

## Minimal reproduction

No proxy, no restarts, no TLS — the "server" is a bare TCP socket that accepts
and stays silent:

```js
import net from 'node:net';
import * as grpc from '@grpc/grpc-js';

// "Server" that accepts TCP connections and never sends HTTP/2 SETTINGS.
net
  .createServer(socket => {
    console.log('[blackhole] TCP connection accepted — staying silent forever');
    socket.on('data', () => {});
    socket.on('error', () => {});
  })
  .listen(50000, '127.0.0.1', () => {
    const client = new grpc.Client(
      '127.0.0.1:50000',
      grpc.credentials.createInsecure(),
      {'grpc.keepalive_time_ms': 30000},
    );

    const startedAt = Date.now();
    console.log('issuing unary call WITHOUT deadline...');
    client.makeUnaryRequest(
      '/test.Test/Get',
      b => b,
      b => b,
      Buffer.alloc(0),
      err => {
        console.log(
          `call finished after ${Math.round(
            (Date.now() - startedAt) / 1000,
          )}s:`,
          err?.message ?? 'ok',
        );
        process.exit(0);
      },
    );

    const states = [
      'IDLE',
      'CONNECTING',
      'READY',
      'TRANSIENT_FAILURE',
      'SHUTDOWN',
    ];
    setInterval(() => {
      console.log(
        `t=+${Math.round((Date.now() - startedAt) / 1000)}s connectivityState=${
          states[client.getChannel().getConnectivityState(false)]
        }`,
      );
    }, 10000);
  });
```

Output:

```
issuing unary call WITHOUT deadline...
[blackhole] TCP connection accepted — staying silent forever
t=+10s connectivityState=CONNECTING
t=+20s connectivityState=CONNECTING   <- expected TRANSIENT_FAILURE per MIN_CONNECT_TIMEOUT
t=+30s connectivityState=CONNECTING
t=+40s connectivityState=CONNECTING
...
```

Expected: the connection attempt fails after ~20 s, the subchannel goes to
`TRANSIENT_FAILURE`, and backoff retries begin. Actual: `CONNECTING` forever,
the call never completes, and the channel never retries.

## Matching the symptoms in this issue

With
`GRPC_TRACE=call_stream,channel,resolving_call,retrying_call,load_balancing_call`,
every new call produces exactly the trace pattern from the original report and
then goes silent:

```
channel            | (1) dns:127.0.0.1:50001 createResolvingCall [9] method="/t.T/Get", deadline=Infinity
resolving_call     | [9] write() called with message of length 0
channel            | (1) dns:127.0.0.1:50001 createRetryingCall [10] method="/t.T/Get"
channel            | (1) dns:127.0.0.1:50001 createLoadBalancingCall [11] method="/t.T/Get"
retrying_call      | [10] write() called with message of length 5
load_balancing_call| [11] write() called with message of length 5
```

No `call_stream` / transport activity follows — the calls are queued waiting for
a subchannel that will never become ready.

The state sequence from this issue (`TRANSIENT_FAILURE → IDLE → CONNECTING`,
with one subchannel briefly reaching `READY`, after which "we stop seeing these
messages altogether") also reproduces if the path only goes silent _after_ a
healthy connection is established. Second reproduction, with a TCP proxy that
starts out healthy, then resets the live connection and silently blackholes
subsequent ones:

```
t=3s state=READY
>>> ghost mode ON: existing flow reset, new connections silently blackholed
state=IDLE
[proxy] ghost: accepted new conn, staying silent
state=CONNECTING
state=CONNECTING
state=CONNECTING
FINAL state=CONNECTING
```

From here the channel never recovers, even after the network path is fully
healthy again for new connections — the hung connection attempt is never
abandoned, so no new attempt is ever made. Only restarting the process (or
recreating the channel) recovers.

## How this happens in practice

This is not a contrived condition. Ours was a real production incident: a
middlebox on the path (LB/NAT) dropped established flows and, for a period,
accepted new TCP connections into a black hole. Two pods out of eight
reconnected during that window and stayed dead for an hour while the other six
recovered normally — they had reconnected slightly outside the window. Server
restarts behind an NLB (as in the original report) can produce the same
half-open-path condition.

## Environment

- `@grpc/grpc-js` **1.14.1** and **1.14.4** (latest at the time of writing) —
  both reproduce
- `master` has the same `createSession()` code, so presumably also affected
- Node.js 22.9.0 (Docker, Debian) and 24.16.0 (macOS) — both reproduce
- No TLS, no proxy, no load balancer needed for the minimal case

## Suggested fix

Arm a timer around the connection attempt in `createSession()` (or in
`Http2SubchannelConnector.connect()`, covering TCP connect + TLS handshake +
SETTINGS), defaulting to `MIN_CONNECT_TIMEOUT` = 20 s per the backoff spec and
ideally configurable via a channel option. On expiry, destroy the session/socket
and reject, so the existing backoff path takes over.

Happy to test a patch against both reproductions.

**Comment by murgatroid99** [comment]
Your analysis and conclusion make sense to me. The omitted `MIN_CONNECT_TIMEOUT` is a holdover from older versions when connection establishment was more completely delegated to the Node http2 module, with the expectation that Node itself would time out unsuccessful connection attempts. Admittedly, I don't know if it ever handled a TCP black hole correctly.

This should be fixed by https://github.com/grpc/grpc-node/pull/3071

**Comment by Sermelyan** [comment]
We hit this failure mode in production too (Node 24, `@grpc/grpc-js` 1.13.4, `round_robin`, a headless Kubernetes Service whose only endpoint IP had moved to another node and was blackholed). The `MIN_CONNECT_TIMEOUT` analysis above matches what we saw, and #3071 addresses the cause of the hang.

I want to flag a **second, independent defect** that this issue's failure mode exposes, because it turns "the channel hangs" into "the process OOMs". It is not fixed by #3071 in the general case, and I could not find an existing issue for it — happy to move this into its own issue if you prefer.

## Queued calls are never removed from `InternalChannel.pickQueue` when they end

`pickQueue` is only ever appended to and drained wholesale. In `internal-channel.ts` (tag `@grpc/grpc-js@1.14.4`) the only push is:

```ts
queueCallForPick(call: LoadBalancingCall) {
  this.pickQueue.push(call);
  this.callRefTimerRef();
}
```

and the only two places it shrinks are the picker update inside `channelControlHelper.updateState`:

```ts
updateState: (connectivityState: ConnectivityState, picker: Picker) => {
  this.currentPicker = picker;
  const queueCopy = this.pickQueue.slice();
  this.pickQueue = [];
  ...
  for (const call of queueCopy) {
    call.doPick();
  }
  this.updateState(connectivityState);
},
```

and `close()`.

When a queued call ends on its own — deadline or cancellation — nothing removes it. `LoadBalancingCall.cancelWithStatus` just funnels into `outputStatus`, which sets `ended = true` and notifies the listener:

```ts
cancelWithStatus(status: Status, details: string): void {
  this.trace('cancelWithStatus code: ' + status + ' details: "' + details + '"');
  this.child?.cancelWithStatus(status, details);
  this.outputStatus(
    { code: status, details: details, metadata: new Metadata() },
    'PROCESSED'
  );
}
```

The call object stays in the array. `doPick()` correctly tolerates this later (`doPick() { if (this.ended) { return; } ... }`), so I assume the intent is "ended entries are harmless, they get skipped at the next drain". That holds only if a picker update is guaranteed to come soon.

## Why that matters here

While a subchannel is stuck in `CONNECTING`, the picker never changes, so `channelControlHelper.updateState` is never called and the drain never runs. Every RPC issued during that window is retained for the lifetime of the channel — and each retained `LoadBalancingCall` pins its `Metadata`, its listener chain, and (through Node's `AsyncLocalStorage` / `AsyncContextFrame`) the caller's entire async context. Correctly setting a deadline on every call does not help: the deadline fires, the caller gets `DEADLINE_EXCEEDED`, and the object stays in the queue.

In one worker after ~10 hours we measured **48,513 `LoadBalancingCall` objects in a single channel's `pickQueue`, retaining ~5.5 GB**.

Two notes on scope:

- This is **not** specific to the stuck-`CONNECTING` bug. Any prolonged period without a picker update has the same effect — e.g. a long name-resolution outage, a slow-starting backend, or `wait_for_ready` calls during an extended `TRANSIENT_FAILURE`. #3071 removes the worst trigger, but the retention itself is still unbounded by anything.
- With `grpc.enable_retries: 1` each `RetryingCall` attempt creates a fresh `LoadBalancingCall`, so the number of retained objects can exceed the number of RPCs the application issued. (I have not measured the multiplier, so treat this part as inference.)

I also could not find any channel option that bounds this — there is no `max_pending_calls` equivalent in `ChannelOptions`, and `grpc.client_idle_timeout_ms` does not help: `enterIdle()` does not touch `pickQueue`, and the idle timer re-arms itself while `callCount > 0`, so it never fires under continuous traffic.

## Prior art: how other gRPC implementations handle this

Every implementation I checked either has no channel-owned queue at all, or removes the call from it when the call ends.

**C-core, current master** — no queue. Each call spawns its own LB-pick promise that loops on a picker observable, in `src/core/client_channel/load_balanced_call_destination.cc`:

```cpp
  unstarted_handler.SpawnGuardedUntilCallCompletes(
      "lb_pick", [unstarted_handler, picker = picker_]() mutable {
        return Map(
            // Wait for the LB picker.
            CheckDelayed(Loop(
                [last_picker = ..., unstarted_handler, picker]() mutable {
                  return Map(
                      picker.Next(last_picker),
                      [...](RefCountedPtr<LoadBalancingPolicy::SubchannelPicker>
                                picker) mutable {
                        ...
                        // Returns 3 possible things:
                        // - Continue to queue the pick
                        // - non-OK status to fail the pick
                        // - a connected subchannel to complete the pick
                        return PickSubchannel(*last_picker, unstarted_handler);
                      });
                })),
```

"Queue the pick" is `Continue{}` in that loop — the waiting state lives in the call's own promise, which is torn down with the call (`SpawnGuardedUntilCallCompletes`).

**Older C-core** did have an array-shaped `lb_queued_calls_` much like `pickQueue`, but it registered a canceller at enqueue time: `OnAddToQueueLocked()` creates an `LbQueuedCallCanceller`, whose constructor does `lb_call_->call_combiner_->SetNotifyOnCancel(&closure_)`, and `CancelLocked` then does `lb_call->RemoveCallFromLbQueuedCallsLocked()` plus `chand->lb_queued_calls_.erase(lb_call)`. Deadline expiry routes through that same path (`deadline_filter.cc` cancels the call combiner), so a queued call that hits its deadline leaves the queue.

**grpc-go** also has no queue: `pickerWrapper` holds one `atomic.Pointer[pickerGeneration]`, a new picker is published by closing the previous generation's channel, and each RPC waits in its own goroutine on `select { case <-ctx.Done(): ...; case <-ch: }`. Deadline or cancellation returns from `pick()` and the whole attempt becomes garbage immediately.

So "a queued call ends by deadline" is treated everywhere else as a case that must release the call. grpc-js currently has the older C-core's queue shape without its cancellation hook.

## Suggested fix

Minimal version: `outputStatus` is the single funnel for every terminal transition, and `LoadBalancingCall` already holds `private readonly channel: InternalChannel`, so the call can deregister itself:

```ts
private outputStatus(status: StatusObject, progress: RpcProgress) {
  if (!this.ended) {
    this.ended = true;
    this.channel.removeCallFromPickQueue(this);   // new
    ...
  }
}
```

To keep that O(1), `pickQueue` could become a `Set<LoadBalancingCall>` instead of an array. `Set` preserves insertion order in JS, so the drain order — and therefore pick fairness — is unchanged, and `close()` can iterate it the same way. The `callRefTimer` invariant ("reffed if and only if `pickQueue` is non-empty") would then also hold accurately rather than only until the next drain.

Structural version, matching what C-core and grpc-go converged on: drop the queue entirely and let each call await the picker itself — the channel keeps only the current picker plus a one-shot "picker changed" signal (a `Promise` or `AbortSignal` in place of C-core's observable and grpc-go's closed channel), and `updateState` swaps the signal. There is then no membership to maintain, so no removal path to forget. That is a larger change, but it removes the bug class rather than one instance of it.

**Comment by murgatroid99** [comment]
I agree that we should remove the call from the pick queue when it ends due to a deadline or cancellation. Not doing that is a simple oversight.

However, that is only a significant problem if is the channel staying in a single state other than READY or IDLE for a long period of time, and that should never happen. The trigger for this specific issue was a channel getting stuck in the CONNECTING state, and that is getting fixed. Any long connection failure should cycle between TRANSIENT_FAILURE and CONNECTING, and the maximum backoff timer length is 20 minutes by default, and configurable. So, if there are in fact other cases where the channel can stay in the same state for a long period of time, those are also bugs that should be fixed.
